AdvLoggerPkg/Library/AdvancedLoggerLib/AdvancedLoggerCommon.c

```diff
@@ -69,20 +69,42 @@ AdvancedLoggerMemoryLoggerWrite (
       if ((UsedSize >= LoggerInfo->LogBufferSize) ||
           ((LoggerInfo->LogBufferSize - UsedSize) < EntrySize))
       {
-        //
-        // Update the number of bytes of log that have not been captured
-        //
-        do {
-          CurrentSize = LoggerInfo->DiscardedSize;
-          NewSize     = CurrentSize + (UINT32)NumberOfBytes;
-          OldSize     = InterlockedCompareExchange32 (
-                          (UINT32 *)&LoggerInfo->DiscardedSize,
-                          (UINT32)CurrentSize,
-                          (UINT32)NewSize
-                          );
-        } while (OldSize != CurrentSize);
-
-        return LoggerInfo;
+        if (FeaturePcdGet (PcdAdvancedLoggerAutoClearEnable) && (LoggerInfo->AtRuntime)) {
+          do {
+            //
+            // Clear the current cursor when auto clear is enabled on buffer full during runtime.
+            //
+            CurrentBuffer = LoggerInfo->LogCurrent;
+
+            NewBuffer     = LoggerInfo->LogBuffer;
+            OldValue      = InterlockedCompareExchange64 (
+                              (UINT64 *)&LoggerInfo->LogCurrent,
+                              (UINT64)CurrentBuffer,
+                              (UINT64)NewBuffer
+                              );
+          } while (OldValue != CurrentBuffer);

```

Comment:
If I am understanding correctly, this is a do-while loop to catch the case where `LogCurrent` is updated between our read and the CmpExch. But, if it was updated, wouldn't we not want to update `LogCurrent` then, because something else updated it, presumably back to `LogBuffer` and we would overwrite new messages?

Reply:
My original thought was to just let the last wrap take precedence. But I think we can leave the re-attempt to the outer loop. So that if this update failed, we can recalculate the pointers and see if the message will just fit in? If not, this logic can engage again?

Reply:
That makes sense to me. If there were a case of very noisy neighbor threads who were spamming the log before this could write in, the platform would need to fix those neighbors or increase the log buffer size, but I think that should be a remote enough edge case we don't need to worry about getting stuck here.

Reply:
Okay, it is updated. Can you please check to see if that does the job to let go the first wrap and keep the log messages?